Tests are failing on the Debian 9 (kernel 4.9.0-13)

The tests are failing: https://github.com/elastio/elastio-snap/runs/1190012942?check_suite_focus=true#step:10:181
```
elastio-snap: 6bba25b
kernel: 4.9.0-13-amd64
gcc: 6.3.0-18+deb9u1)
bash: 4.4.12(1)-release
python: Python 3.5.3

test_destroy_active_incremental (test_destroy.TestDestroy) ... FAIL
test_destroy_active_snapshot (test_destroy.TestDestroy) ... FAIL
test_destroy_dormant_incremental (test_destroy.TestDestroy) ... skipped 'Broken since 4.17 (see #144)'
test_destroy_dormant_snapshot (test_destroy.TestDestroy) ... skipped 'Broken since 4.17 (see #144)'
test_destroy_nonexistent_device (test_destroy.TestDestroy) ... FAIL
test_destroy_unverified_incremental (test_destroy.TestDestroy) ... umount: /tmp/elastio-snap: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
ERROR
test_destroy_unverified_snapshot (test_destroy.TestDestroy) ... umount: /tmp/elastio-snap: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
umount: /tmp/elastio-snap: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
ERROR
ERROR
ERROR
ERROR
ERROR

======================================================================
ERROR: test_destroy_unverified_incremental (test_destroy.TestDestroy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_destroy.py", line 81, in test_destroy_unverified_incremental
    util.unmount(self.mount)
  File "/home/elastio/elastio-snap/tests/util.py", line 23, in unmount
    subprocess.check_call(cmd, timeout=10)
  File "/usr/lib/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['umount', '/tmp/elastio-snap']' returned non-zero exit status 32

======================================================================
ERROR: test_destroy_unverified_snapshot (test_destroy.TestDestroy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_destroy.py", line 72, in test_destroy_unverified_snapshot
    util.unmount(self.mount)
  File "/home/elastio/elastio-snap/tests/util.py", line 23, in unmount
    subprocess.check_call(cmd, timeout=10)
  File "/usr/lib/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['umount', '/tmp/elastio-snap']' returned non-zero exit status 32

======================================================================
ERROR: tearDownClass (test_destroy.TestDestroy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/devicetestcase.py", line 34, in tearDownClass
    util.unmount(cls.mount)
  File "/home/elastio/elastio-snap/tests/util.py", line 23, in unmount
    subprocess.check_call(cmd, timeout=10)
  File "/usr/lib/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['umount', '/tmp/elastio-snap']' returned non-zero exit status 32

======================================================================
ERROR: setUpClass (test_setup.TestSetup)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/devicetestcase.py", line 24, in setUpClass
    cls.kmod.load(debug=1)
  File "/home/elastio/elastio-snap/tests/kmod.py", line 31, in load
    timeout=self.timeout)
  File "/usr/lib/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['insmod', '../src/elastio-snap.ko', 'debug=1']' returned non-zero exit status 1

======================================================================
ERROR: setUpClass (test_snapshot.TestSnapshot)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/devicetestcase.py", line 24, in setUpClass
    cls.kmod.load(debug=1)
  File "/home/elastio/elastio-snap/tests/kmod.py", line 31, in load
    timeout=self.timeout)
  File "/usr/lib/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['insmod', '../src/elastio-snap.ko', 'debug=1']' returned non-zero exit status 1

======================================================================
ERROR: setUpClass (test_transition_incremental.TestTransitionToIncremental)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/devicetestcase.py", line 24, in setUpClass
    cls.kmod.load(debug=1)
  File "/home/elastio/elastio-snap/tests/kmod.py", line 31, in load
    timeout=self.timeout)
  File "/usr/lib/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['insmod', '../src/elastio-snap.ko', 'debug=1']' returned non-zero exit status 1

======================================================================
FAIL: test_destroy_active_incremental (test_destroy.TestDestroy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_destroy.py", line 40, in test_destroy_active_incremental
    self.assertEqual(elastio_snap.transition_to_incremental(self.minor), 0)
AssertionError: 5 != 0

======================================================================
FAIL: test_destroy_active_snapshot (test_destroy.TestDestroy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_destroy.py", line 32, in test_destroy_active_snapshot
    self.assertEqual(elastio_snap.setup(self.minor, self.device, self.cow_full_path), 0)
AssertionError: 16 != 0

======================================================================
FAIL: test_destroy_nonexistent_device (test_destroy.TestDestroy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_destroy.py", line 29, in test_destroy_nonexistent_device
    self.assertEqual(elastio_snap.destroy(self.minor), errno.ENOENT)
AssertionError: 0 != 2

----------------------------------------------------------------------
Ran 7 tests in 1.150s

FAILED (failures=3, errors=6, skipped=2)
Error: Process completed with exit code 1.

```
Or even hanging: https://github.com/elastio/elastio-snap/runs/1189842448?check_suite_focus=true#step:10:184
```
elastio-snap: fedecce
kernel: 4.9.0-13-amd64
gcc: 6.3.0-18+deb9u1)
bash: 4.4.12(1)-release
python: Python 3.5.3
test_destroy_active_incremental (test_destroy.TestDestroy) ... FAIL

test_destroy_active_snapshot (test_destroy.TestDestroy) ... FAIL
test_destroy_dormant_incremental (test_destroy.TestDestroy) ... skipped 'Broken since 4.17 (see #144)'
test_destroy_dormant_snapshot (test_destroy.TestDestroy) ... skipped 'Broken since 4.17 (see #144)'
Error: The operation was canceled.
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -63,9 +63,6 @@ jobs:
         working-directory: ${{env.BOX_DIR}}
 
       - name: Run tests
-        # TODO: Bring back tests for Debian 9 when issue https://github.com/elastio/elastio-snap/issues/50
-        # is fixed
-        if: ${{ matrix.distro != 'debian9' }}
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c 'cd tests && sudo ./elio-test.sh'
         working-directory: ${{env.BOX_DIR}}
         # For now tests are taking 10-20 seconds. But they can hang.

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -214,6 +214,7 @@ static struct block_device *blkdev_get_by_path(const char *path, fmode_t mode, v
 #endif
 
 #ifndef HAVE_SUBMIT_BIO_1
+//#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0)
 
 #ifndef REQ_DISCARD
 #define REQ_DISCARD 0
@@ -738,8 +739,20 @@ static inline void elastio_snap_bio_copy_dev(struct bio *dst, struct bio *src){
 #define BIO_SET_SIZE 256
 #define bio_last_sector(bio) (bio_sector(bio) + (bio_size(bio) / SECTOR_SIZE))
 
-// As of Linux 5.2, __REQ_NR_BITS == 26
-#define __ELASTIO_SNAP_PASSTHROUGH 30    /* don't perform COW operation */
+/* don't perform COW operation */
+#ifdef HAVE_ENUM_REQ_OP
+//#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0) && LINUX_VERSION_CODE < KERNEL_VERSION(4,10,0)
+/* special case for deb9's 4.9 train
+ * Bit 30 conflicts with struct bio's bi_opf opcode bitfield, which occupies the top 3 bits of the member. If we set
+ * that bit, it will mutate the operation that the bio is representing. Setting this to 28 puts this in an unused flag
+ * for bi_opf (that flag means something in struct request's cmd_flags, but we're not setting that).
+ */
+#define __ELASTIO_SNAP_PASSTHROUGH 28	// set as the last flag bit
+#else
+// set as an unused flag in versions older than 4.8
+// set as an unused opcode bit in kernels newer than 4.9
+#define __ELASTIO_SNAP_PASSTHROUGH 30
+#endif
 #define ELASTIO_SNAP_PASSTHROUGH (1ULL << __ELASTIO_SNAP_PASSTHROUGH)
 
 #define ELASTIO_SNAP_DEFAULT_SNAP_DEVICES 24

--- a/src/elastio-snap.h
+++ b/src/elastio-snap.h
@@ -15,7 +15,7 @@
 #include <linux/ioctl.h>
 #include <linux/limits.h>
 
-#define ELASTIO_SNAP_VERSION "0.10.13"
+#define ELASTIO_SNAP_VERSION "0.10.14"
 #define ELASTIO_IOCTL_MAGIC 'A' // 0x41
 
 struct setup_params{

--- a/tests/elio-test.sh
+++ b/tests/elio-test.sh
@@ -15,7 +15,7 @@ packman="apt-get"
 which yum >/dev/null && packman=yum
 which pip3 >/dev/null || $packman install -y python3-pip
 
-if ! pip3 list | grep -q cffi ; then
+if ! pip3 list --format=legacy 2>/dev/null | grep -q cffi ; then
     echo "Python module CFFI is not installed. Installing it..."
     pip3 install cffi
 fi

